subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/AbstractInputFilePropertyAnnotationHandler.java

```diff
@@ -19,24 +19,27 @@
 import org.gradle.api.internal.tasks.DeclaredTaskInputFileProperty;
 import org.gradle.api.internal.tasks.PropertySpecFactory;
 import org.gradle.api.internal.tasks.TaskValidationContext;
+import org.gradle.api.internal.tasks.ValidatingValue;
 import org.gradle.api.internal.tasks.properties.BeanPropertyContext;
-import org.gradle.api.internal.tasks.properties.PropertyValue;
 import org.gradle.api.internal.tasks.properties.PropertyVisitor;
 import org.gradle.api.tasks.PathSensitive;
 import org.gradle.api.tasks.PathSensitivity;
 import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.internal.reflect.PropertyMetadata;
 
 import java.io.File;
 
-public abstract class AbstractInputPropertyAnnotationHandler implements PropertyAnnotationHandler {
+import static org.gradle.api.internal.tasks.properties.annotations.InputPropertyAnnotationHandlerUtils.isOptional;
+
+public abstract class AbstractInputFilePropertyAnnotationHandler implements PropertyAnnotationHandler {
     @Override
     public boolean shouldVisit(PropertyVisitor visitor) {
         return !visitor.visitOutputFilePropertiesOnly();
```

Comment:
IIRC, that is a performance optimization introduced by @adammurdoch so we don't have a performance regression for attaching the producers. The performance regression was caused by going into the nested properties. See https://github.com/gradle/gradle/blob/69efd80d417b5d79508206b91ad59339f2f88cc0/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/PropertyAssociationTaskFactory.java#L52-L88

It would be better to be able to attach the producer to outputs in nested beans as well. I'll check again how big the regression is.

Reply:
👍

Reply:
I checked it out, and there seems to be quite a performance impact: https://builds.gradle.org/viewLog.html?buildId=19160450&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance&branch_Gradle_Check_Stage_ReadyforMerge=wolfs%2Fextract-property-specs-review

I would leave the code as is for now. Shall we add a card to clean it up?

Reply:
Yes please.

Reply:
https://github.com/gradle/build-cache/issues/1438